File: bully-election/BullyElection.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Scanner;

/**
 * Simulation of the Bully election algorithm for coordinator selection.
 * The coordinator is the active process with the highest ID.
 * When the coordinator fails, any process can start an election.
 */
public class BullyElection {

    static class Process {
        final int id;
        boolean alive;

        Process(int id) {
            this.id = id;
            this.alive = true;
        }

        @Override
        public String toString() {
            return "P" + id + (alive ? "" : " (dead)");
        }
    }

    private final List<Process> processes;
    private int coordinatorId;

    public BullyElection(List<Process> processes) {
        this.processes = new ArrayList<>(processes);
        this.coordinatorId = findHighestAliveId();
    }

    private int findHighestAliveId() {
        return processes.stream()
                .filter(p -> p.alive)
                .mapToInt(p -> p.id)
                .max()
                .orElse(-1);
    }

    /** Simulate coordinator failure and run bully election from starter process. */
    public void failCoordinator() {
        Process coord = processes.stream()
                .filter(p -> p.id == coordinatorId)
                .findFirst()
                .orElse(null);
        if (coord != null) {
            System.out.println("\n>>> Coordinator P" + coordinatorId + " FAILED.");
            coord.alive = false;
        }
        coordinatorId = -1;
    }

    /**
     * Bully election: process {@code starter} sends ELECTION to all higher IDs.
     * If any higher alive process exists, it will take over (recursive simulation).
     * Otherwise starter becomes coordinator and sends OK to lower processes.
     */
    public void runElectionFrom(int starterId) {
        Process starter = processes.stream()
                .filter(p -> p.id == starterId)
                .findFirst()
                .orElse(null);
        if (starter == null || !starter.alive) {
            System.out.println("Starter P" + starterId + " is not active; cannot start election.");
            return;
        }

        System.out.println("\n--- Election started by P" + starterId + " ---");

        List<Process> higher = new ArrayList<>();
        for (Process p : processes) {
            if (p.alive && p.id > starterId) {
                higher.add(p);
            }
        }
        higher.sort(Comparator.comparingInt(p -> p.id));

        if (higher.isEmpty()) {
            coordinatorId = starterId;
            System.out.println("No higher-ID alive processes. P" + starterId + " becomes COORDINATOR.");
            announceCoordinator(starterId);
            return;
        }

        System.out.println("P" + starterId + " sends ELECTION to: " + formatIds(higher));
        // Highest alive respondent wins bully election
        Process winner = higher.get(higher.size() - 1);
        System.out.println("Highest respondent P" + winner.id + " takes over election.");
        runElectionFrom(winner.id);
    }

    private void announceCoordinator(int id) {
        List<Integer> lower = new ArrayList<>();
        for (Process p : processes) {
            if (p.alive && p.id < id) {
                lower.add(p.id);
            }
        }
        if (!lower.isEmpty()) {
            System.out.println("P" + id + " sends COORDINATOR message to: " + lower);
        }
    }

    private static String formatIds(List<Process> list) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < list.size(); i++) {
            if (i > 0) {
                sb.append(", ");
            }
            sb.append("P").append(list.get(i).id);
        }
        return sb.toString();
    }

    public void printState() {
        System.out.println("\n=== Cluster state ===");
        List<Process> sorted = new ArrayList<>(processes);
        sorted.sort(Comparator.comparingInt(p -> p.id));
        for (Process p : sorted) {
            System.out.println("  " + p);
        }
        if (coordinatorId >= 0) {
            Process c = processes.stream().filter(x -> x.id == coordinatorId).findFirst().orElse(null);
            if (c != null && c.alive) {
                System.out.println("Current coordinator: P" + coordinatorId);
            } else {
                System.out.println("No coordinator (election needed).");
            }
        } else {
            System.out.println("No coordinator (election needed).");
        }
    }

    public int getCoordinatorId() {
        return coordinatorId;
    }

    public static void main(String[] args) {
        List<Process> procs = new ArrayList<>();
        procs.add(new Process(1));
        procs.add(new Process(2));
        procs.add(new Process(3));
        procs.add(new Process(4));
        procs.add(new Process(5));

        BullyElection sim = new BullyElection(procs);
        System.out.println("Bully Election Algorithm — simulation");
        sim.printState();

        Scanner sc = new Scanner(System.in);
        while (true) {
            System.out.println("\nCommands: 1=fail coordinator & elect from P1, 2=elect from custom ID, 3=state, q=quit");
            System.out.print("> ");
            if (!sc.hasNextLine()) {
                break;
            }
            String line = sc.nextLine().trim();
            if (line.equalsIgnoreCase("q")) {
                break;
            }
            if (line.equals("1")) {
                sim.failCoordinator();
                sim.printState();
                int starter = 1;
                Process p1 = procs.stream().filter(p -> p.id == starter).findFirst().orElse(null);
                if (p1 != null && p1.alive) {
                    sim.runElectionFrom(starter);
                    sim.printState();
                } else {
                    // pick lowest alive
                    int low = procs.stream().filter(p -> p.alive).mapToInt(p -> p.id).min().orElse(-1);
                    if (low >= 0) {
                        sim.runElectionFrom(low);
                        sim.printState();
                    }
                }
            } else if (line.equals("2")) {
                System.out.print("Enter starter process ID: ");
                if (sc.hasNextInt()) {
                    int sid = sc.nextInt();
                    sc.nextLine();
                    sim.runElectionFrom(sid);
                    sim.printState();
                } else {
                    sc.nextLine();
                    System.out.println("Invalid ID.");
                }
            } else if (line.equals("3")) {
                sim.printState();
            } else if (line.equalsIgnoreCase("demo")) {
                sim.printState();
                sim.failCoordinator();
                sim.printState();
                sim.runElectionFrom(1);
                sim.printState();
            } else {
                System.out.println("Unknown command. Try: demo (non-interactive full demo)");
            }
        }
        sc.close();
        System.out.println("Bye.");
    }
}

File: bully-election/bully_election.c
/*
 * Bully election algorithm — simplified C simulation.
 * Coordinator = highest ID among alive processes.
 * Compile: gcc -o bully bully_election.c -std=c99
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_P 16

typedef struct {
    int id;
    int alive;
} Process;

static Process processes[MAX_P];
static int nproc;
static int coordinator = -1;

static int find_highest_alive(void) {
    int best = -1;
    for (int i = 0; i < nproc; i++) {
        if (processes[i].alive && processes[i].id > best) {
            best = processes[i].id;
        }
    }
    return best;
}

static void print_state(void) {
    printf("\n=== Cluster state ===\n");
    for (int i = 0; i < nproc; i++) {
        printf("  P%d %s\n", processes[i].id,
               processes[i].alive ? "" : "(dead)");
    }
    if (coordinator >= 0) {
        int ok = 0;
        for (int i = 0; i < nproc; i++) {
            if (processes[i].id == coordinator && processes[i].alive) {
                ok = 1;
                break;
            }
        }
        if (ok) {
            printf("Current coordinator: P%d\n", coordinator);
        } else {
            printf("No coordinator (election needed).\n");
        }
    } else {
        printf("No coordinator (election needed).\n");
    }
}

static void run_election_from(int starter_id) {
    int si = -1;
    for (int i = 0; i < nproc; i++) {
        if (processes[i].id == starter_id) {
            si = i;
            break;
        }
    }
    if (si < 0 || !processes[si].alive) {
        printf("Starter P%d is not active.\n", starter_id);
        return;
    }

    printf("\n--- Election started by P%d ---\n", starter_id);

    int higher[MAX_P];
    int nh = 0;
    for (int i = 0; i < nproc; i++) {
        if (processes[i].alive && processes[i].id > starter_id) {
            higher[nh++] = processes[i].id;
        }
    }

    if (nh == 0) {
        coordinator = starter_id;
        printf("No higher-ID alive processes. P%d becomes COORDINATOR.\n", starter_id);
        return;
    }

    printf("P%d sends ELECTION to higher processes: ", starter_id);
    for (int i = 0; i < nh; i++) {
        printf("P%d%s", higher[i], (i < nh - 1) ? ", " : "\n");
    }

    int winner = higher[nh - 1];
    printf("Highest respondent P%d takes over election.\n", winner);
    run_election_from(winner);
}

static void fail_coordinator(void) {
    for (int i = 0; i < nproc; i++) {
        if (processes[i].id == coordinator) {
            printf("\n>>> Coordinator P%d FAILED.\n", coordinator);
            processes[i].alive = 0;
            break;
        }
    }
    coordinator = -1;
}

int main(void) {
    /* Default: processes 1..5 */
    nproc = 5;
    for (int i = 0; i < nproc; i++) {
        processes[i].id = i + 1;
        processes[i].alive = 1;
    }
    coordinator = find_highest_alive();

    printf("Bully Election Algorithm — C simulation\n");
    print_state();

    printf("\nDemo: fail coordinator, then elect from P1\n");
    fail_coordinator();
    print_state();
    run_election_from(1);
    print_state();

    return 0;
}
